### ISSUE TITLE
try-catch-else

### ISSUE THREAD

**Issue by christopher-dG** [JuliaLang/julia#21130]
Coming from Python, I was really surprised to find that Julia has no way to execute code after a `try` block _only_ if no exception/error was caught. Right now it seems the only way to get around this is something like:
```julia
try
  maybe_raise_exception()
  unrelated_stuff()
  more_unrelated()
catch
  stuff()
end 
```
Whereas the Python code for this would read:
```python
try:
  maybe_raise_exception()
except SomeException:
  stuff()
else:
  unrelated_stuff()
  more_unrelated() 
```
It seems like bad practice to put code that we aren't worried about causing exceptions/errors in a `try` block so I'm interested in hearing what others have to say about this.

e: Alternatively you could do:
```julia
try
  maybe_raise_exception()
  noex = true
catch
  stuff()
end
if noex
  unrelated_stuff()
  more_unrelated()
end
```
Although I do personally prefer the `else` syntax.

**Comment by JeffBezanson** [comment]
Yes, this seems like a good feature.

**Comment by yuyichao** [comment]
Ref https://github.com/JuliaLang/julia/issues/1289 which is somewhat more confusing than `try-catch-else`....

**Comment by stevengj** [comment]
I don't see the problem with
```
try
  maybe_raise_exception()
  unrelated_stuff()
  more_unrelated()
catch
  stuff()
end 
```
There's no code duplication, and it seems like this is a normal way to use a `try` block: the whole block executes only if there is no exception.

**Comment by yuyichao** [comment]
I think the issue is that `unrelated_stuff()` and `more_unrelated()` can throw exceptions that you don't want to catch since they are "unrelated".

**Comment by christopher-dG** [comment]
What @yuyichao said. Additionally, I don't think that code that has no risk of throwing an exception should be in a `try` block.

**Comment by StefanKarpinski** [comment]
See also https://github.com/JuliaLang/julia/issues/7026. The whole reason for try-catch-else is so that you don't accidentally catch an unexpected error thrown by `unrelated_stuff` that just happens to look like the expected error thrown by `maybe_raise_exception`. In the chain of custody design, that's not possible since you'd need to annotate `maybe_raise_exception` with `throws SomeExceptionType` in order to be able to catch it the exception it throws. It seems to me that this is a feature that just sweeps the real problem of catching the wrong exception under the carpet.

**Comment by JeffBezanson** [comment]
I think this is a reasonable feature idea. I can see how it's somewhat redundant with the chain-of-custody design, but handles a much simpler and more limited case. It's not really necessary to be able to catch exceptions with fine granularity when all you need is to pop the exception handler for an easily-identified block of code.

**Comment by christopher-dG** [comment]
Sure, I was just closing due to lack of activity. Maybe we can see it in the future :slightly_smiling_face:.

**Comment by JeffBezanson** [comment]
Yes, there have definitely been cases where we implemented a feature literally years after it was requested. This one might not even take that long :-P

**Comment by heetbeet** [comment]
I developed this for myself to get around some of the try/catch problems I encountered:
https://github.com/heetbeet/TryCatch.jl

**Comment by StefanKarpinski** [comment]
There doesn't seem to be any real objection to this feature, someone just has to implement it. I've occasionally wanted this.

**Comment by StefanKarpinski** [comment]
I've removed the design and speculative labels since there doesn't seem to be much to design: allow an `else` block on a `try`-`catch` that is evaluated if nothing is caught. Doesn't seem especially speculative either.
